### ISSUE TITLE
Cannot split pcap with -C into files bigger than 2GB

### ISSUE THREAD

**Issue by jmlich** [the-tcpdump-group/tcpdump#488]
Hi I am trying to split pcap into multiple files with size 100GB with following command:

```
tcpdump -r ./network.pcap -C 100000 -w ./output
```

It seems this doesn't work becuase ndo_Cflag (./netdissect.h and ./interface.h) is defined as int

```
int ndo_Cflag;                /* rotate dump files after this many bytes */
```

The value if Cflag is given by following formula (main.c):

```
Cflag = atoi(optarg) * 1000000;
```

**Comment by mcr** [comment]
Jozef Mlich <notifications@github.com> wrote:
    > Hi I am trying to split pcap into multiple files with size 100GB with
    > following command:

    > tcpdump -r ./network.pcap -C 100000 -w ./output

    > It seems this doesn't work becuase ndo_Cflag (./netdissect.h and ./
    > interface.h) is defined as int

    > int ndo_Cflag; /* rotate dump files after this many bytes */

    > The value if Cflag is given by following formula (main.c):

    > Cflag = atoi(optarg) * 1000000;

So, you are saying we need to use a 64-bit value here, and maybe
a string conversion function that returns a bigger type.

A tested patch would be most appreciated...

--
]               Never tell me the odds!                 | ipv6 mesh networks [
]   Michael Richardson, Sandelman Software Works        | network architect  [
]     mcr@sandelman.ca  http://www.sandelman.ca/        |   ruby on rails    [

**Comment by stevekay** [comment]
Sounds like a switch from int to unsigned long long and atoi to strtoull would address this then ?

    [steve@fedora tcpdump]$ git diff
    diff --git a/tcpdump.c b/tcpdump.c
    index e023706..44701f9 100644
    --- a/tcpdump.c
    +++ b/tcpdump.c
    @@ -127,7 +127,7 @@ The Regents of the University of California.  All rights reserved.\n";
     #define SIGNAL_REQ_INFO SIGUSR1
     #endif
    
    -static int Cflag;                      /* rotate dump files after this many bytes */
    +static unsigned long long Cflag;        /* rotate dump files after this many bytes */
     static int Cflag_count;                        /* Keep track of which file number we're writing */
     static int Dflag;                      /* list available devices and exit */
     /*
    @@ -864,7 +864,7 @@ main(int argc, char **argv)
                            break;
    
                    case 'C':
    -                       Cflag = atoi(optarg) * 1000000;
    +                       Cflag = strtoull(optarg,NULL,10) * 1000000;
                            if (Cflag < 0)
                                    error("invalid file size %s", optarg);
                            break;

**Comment by stevekay** [comment]
Fix would be along the lines of Cflag int -> unsigned long long, atoi -> strtoull ?

    [steve@fedora tcpdump]$ git diff
    diff --git a/tcpdump.c b/tcpdump.c
    index e023706..44701f9 100644
    --- a/tcpdump.c
    +++ b/tcpdump.c
    @@ -127,7 +127,7 @@ The Regents of the University of California.  All rights reserved.\n";
     #define SIGNAL_REQ_INFO SIGUSR1
     #endif
    
    -static int Cflag;                      /* rotate dump files after this many bytes */
    +static unsigned long long Cflag;        /* rotate dump files after this many bytes */
     static int Cflag_count;                        /* Keep track of which file number we're writing */
     static int Dflag;                      /* list available devices and exit */
     /*
    @@ -864,7 +864,7 @@ main(int argc, char **argv)
                            break;
    
                    case 'C':
    -                       Cflag = atoi(optarg) * 1000000;
    +                       Cflag = strtoull(optarg,NULL,10) * 1000000;
                            if (Cflag < 0)
                                    error("invalid file size %s", optarg);
                            break;
    [steve@fedora tcpdump]$

**Comment by patrickj0607** [comment]
@stevekay Thanks a lot for this patch! As I needed this working, I just cloned the repo, applied your patch and compiled it - it works perfectly.

**Comment by infrastation** [comment]
Commit 9868446 should fix the problem in a more straightforward way, please reopen the issue if it does not. Thank you for the patience.

**Comment by guyharris** [comment]
This should fix the problem on 64-bit UN\*Xes, which probably handles most of the cases where it's an issue.  (We'd need to change the libpcap API to include a routine that always returns a 64-bit file offset, and make `Cflag` a 64-bit variable, to fix it on 32-bit UN\*Xes and on Windows.)

**Comment by infrastation** [comment]
Yep.

**Comment by martin-kieliszek** [comment]
Hi @infrastation,

I am on Redhat 7.2 x86_64
I have pulled libpcap-1.8.1
I have pulled tcpdump-4.9.2 (and confirmed that tcpdump.h has the static long Cflag modification)

I have 1) installed flex and bison in order to ./configure make and make install libpcap
I have 2) ./configure make and make install tcpdump-4.9.2

At the completion of the make install --> tcpdump and tcpdump.4.9.2 has been placed in /usr/local/sbin/

Verifying version
```
# /usr/local/sbin/tcpdump.4.9.2 --version
tcpdump.4.9.2 version 4.9.2
libpcap version 1.8.1

# /usr/local/sbin/tcpdump --version
tcpdump version 4.9.2
libpcap version 1.8.1
```

Now, when attempting to set a capture size > 2000MB, I am still encountering an error.
**Note my agenda: I want to capture one file up to 200,000 MB and am currently limited to 2,200MB despite trying to update tcpdump to the fixed version you have proposed**

I apologise in advance for bringing this up if I have made an obvious mistake but can you alleviate my confusion?

See below for errors

```
# /usr/local/sbin/tcpdump -i eth0 -W 1 -C 2200
tcpdump: invalid file size 2200

# /usr/local/sbin/tcpdump -i eth0 -W -C 2000
listening on eth0, link-type ....... <-- works fine with -C set to 2000


# /usr/local/sbin/tcpdump.4.9.2 -i eth0 -W 1 -C 2200
tcpdump.4.9.2: invalid file size 2200

# /usr/local/sbin/tcpdump.4.9.2 -i eth0 -W -C 2000
listening on eth0, link-type ....... <-- works fine with -C set to 2000
```

**Comment by guyharris** [comment]
> I have pulled tcpdump-4.9.2 (and confirmed that tcpdump.c has the static long Cflag modification)

That's not sufficient to fix this problem!  It's set from `atoi() * 1000000`, which is two `int` values multiplied together, which means the result is an `int`, and it will overflow.

**Comment by guyharris** [comment]
OK, eff1d1265654068c3c2408a571173f72e82e8a2d should fix that.  @infrastation, should we backport it to 4.9.x?

**Comment by infrastation** [comment]
If it works this time then definitely yes because this bug is claimed as fixed in the 4.9.1 change log section. As to the previous "bugfix" not actually fixing the issue, I've been trying to recall today what it actually was that I vaguely remember testing for >2GB file support and seeing it work, if that was not this issue. Thanks for the follow-ups folks.
